### ISSUE TITLE
No report of implicit any type in a generator when destructuring 

### ISSUE THREAD

**Issue by Igorbek** [microsoft/TypeScript#43172]
# Bug Report

### 🔎 Search Terms

no implicit any, generator, yield, destructure
#41348 - Report implicit any error for 'yield' result with no contextual type

### 🕗 Version & Regression Information

TypeScript 4.2.3

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?#code/GYVwdgxgLglg9mAVAAgM4EMC2AHANgUwAoBKZAbwChlrkIFUpl1kBeZATxn1wBNkBGZAHohyfACdxccQBomqMQA9s+aPh5zmMBTBy4YEGFFzsmYdlRp0wDcsgBGyAL6sOXXgOGiwcMZOly9iCM+MqqUOqByNrRegZGJmYWNLT0jADaEAC6rpzcfIIiyD5+UrIOwUoqahq00Tpxhsam6OYUThRAA)

### 💻 Code

```ts
function* sample() {
    const a = yield 1 // error, as expected, a is implicitly any
    const { b } = yield 1 // no error, but expected, b is implicitly any
    const [c] = yield 1 // no error, but expected, c is implicitly any
}
```

### 🙁 Actual behavior

`b`, `c` implicitly have type `any`, but no error reported

### 🙂 Expected behavior

Errors reported

**Comment by a-tarasyuk** [comment]
@RyanCavanaugh What about the following case? 

```ts
function* g() {
    const [a = 1, b = 2] = yield; ???
}
```

**Comment by Igorbek** [comment]
@a-tarasyuk  it seems they're given context types, so it should not be an error. However, the intention might be:
```
const [a = 1, b = 2]: [number | string, number | boolean] = yield;
```
not really definitive.

**Comment by a-tarasyuk** [comment]
Should the expected behavior be like this?

```ts
function* g1() {
    const { a } = yield;                 // error: implicit any
    const { b }: { b: string } = yield;  // ok
    const { c = 1 } = yield;             // ok ??
}

function* g2() {
    const [a] = yield;           // error: implicit any
    const [b]: [string] = yield; // ok
    const [c = 1] = yield;       //  ok ??
}
```

/cc @DanielRosenwasser @RyanCavanaugh

**Comment by Igorbek** [comment]
I think you're correct with inferred type from the default being ok, as it'd be consistent with plain functions:
```ts
function f2({ a }) {} // error
function f2({ a }: { a: number }) {} // ok
function f3({a = 1}) {} // ok
```

**Comment by Igorbek** [comment]
ah, on the second thought, if it is `const` binding, the type is not widening, which is not the same in the argument destructuring. You either end up having a type of const context or have inconsistency and widen the type even in `const` binding.
I'd say in this case it should be an error then. Although `let` would work just fine, for consistency, I'd also vote for the error.
